src/UserSpaceInstrumentation/include/UserSpaceInstrumentation/RegisterState.h

```diff
@@ -0,0 +1,240 @@
+// data needs to call the "Has*DataStored" functions first.
+//
+// The fpu, mmx, sse registers can be accessed via the "GetFxSave" call using the structs defined
+// above. Note that the fpu and the mmx registers share the same memory since their usage is
+// mutually exclusive. The upper half of the avx registers can be accessed via "GetAvxHiRegisters"
+// (low half is shared with sse registers).
+//
+// There are more additional registers that get backed up and restored by this class (if the cpu
+// supports them and backup is necessary) but no accessors are defined to alter that data (e.g.
+// Avx-512, Mpx registers, ...). It should be straightforward to add support for these if that
+// should become necessary. Details again can be found in [1].
+//
+// [1]: "Intel 64 and IA-32 Architectures Software Developer’s Manual, Volume 1", section 13
+// [2]: https://www.moritz.systems/blog/how-debuggers-work-getting-and-setting-x86-registers-part-2/
+//
+// Example usage:
+//
+//    // Wait for some thread to be halted.
+//    waitpid(pid, &status, 0);
+//
+//    // Alter Avx state if applicable.
+//    RegisterState s;
+//    auto result_backup = s.BackupRegisters(pid);
+//    if (s.HasAvxDataStored()){
+//      s.GetAvxHiRegisters()->ymm[0].bytes[0] = 42;
+//    }
+//    auto result_restore = s.RestoreRegisters();
+//
+//    // Continue thread with altered state.
+//    ptrace(PT_CONTINUE, pid, 1, 0);
+class RegisterState {
+ public:
+  [[nodiscard]] ErrorMessageOr<void> BackupRegisters(pid_t tid);
+  [[nodiscard]] ErrorMessageOr<void> RestoreRegisters();
+
+  enum class Bitness {
+    k32Bit,
+    k64Bit,
+  };
+  Bitness GetBitness() const { return bitness_; }
+
+  // Returns a pointer to a union of 32 and 64 bit general purpose registers. Call "GetBitness" to
+  // determine which member of the union is valid.
+  GeneralPurposeRegisters* GetGeneralPurposeRegisters() { return &general_purpose_registers_; }
+
+  // Some registers do not get stored in RegisterState. The Cpu might not support them or they might
+  // be in their initial state. So before accessing this data one needs to call the "Has*DataStored"
+  // methods first.
+  bool Hasx87DataStored();
+  bool HasSseDataStored();
+  bool HasAvxDataStored();
+
+  // Structure access to the different parts of the XSave area. Detail, again, can be found at
+  // "Intel 64 and IA-32 Architectures Software Developer’s Manual, Volume 1", section 13.
+  // "GetFxSave" can be used to access fpu, mmx, sse registers. "GetAvxHiRegisters" give access to
+  // the upper half of the avx registers (lower half in store in the sse registers).
+  FxSave* GetFxSave() { return reinterpret_cast<FxSave*>(xsave_area_.data()); }
+  XSaveHeader* GetXSaveHeader() { return reinterpret_cast<XSaveHeader*>(xsave_area_.data() + 512); }
+  YmmHi* GetAvxHiRegisters() { return reinterpret_cast<YmmHi*>(xsave_area_.data() + avx_offset_); }
```

Comment:
`[[nodiscard]]` here too, since you use it above and these are accessors?

Reply:
done